Make ssh X-forward password typing robust

okurz opened the pull request:
- Add additional needle checks to accept the host key if it was not already
   accepted in before
 - Make sure the password prompt is shown before starting to type

 - Make sure the second xterm window is shown not to yield a false negative
matching on the test pattern blindly typed into the first window when no actual
ssh connection could be established

Fixes https://progress.opensuse.org/issues/5346

screenshot of host key authentication needle matching:
![openqa_sshxterm_hostkey_authentication_prompt](https://cloud.githubusercontent.com/assets/1693432/12551761/0425b0b8-c36d-11e5-991c-98b4b69e2085.png)

screenshot of password prompt needle matching:
![openqa_sshxterm_password_prompt](https://cloud.githubusercontent.com/assets/1693432/12551763/04427658-c36d-11e5-8747-359e430856ce.png)

local verification of sshxterm:
![openqa_sshxterm_verification](https://cloud.githubusercontent.com/assets/1693432/12551764/0445671e-c36d-11e5-94c5-88ea263adc58.png)

dzedro commented on tests/x11/sshxterm.pm:
```diff
@@ -16,19 +16,22 @@ sub run() {
+    # close both windows
     send_key "alt-f4";
-    sleep 1;
     send_key "alt-f4";
```
I guess one window will be not closed under higher load, but maybe I'm getting openQA paranoid :)

okurz replied:
yes, I was fearing the same. So maybe we actually need another assert_screen here. But basically we are already in the "cleanup" phase. Any other idea how we could handle this safely? And more important: How can I test this locally :-)

coolo replied:
I don't think assert_screen is required. But wait_still_screen and wait_screen_change are both valid replacements for that sleep

okurz replied:
I am not sure on the semantics here, should it be
```
send_key "alt-f4";
wait_screen_change {
    send_key "alt-f4";
}
```
?

coolo replied:
I was more thinking along 
```
wait_screen_change {
    send_key "alt-f4";
}
send_key "alt-f4";
```

okurz replied:
On Monday 25 January 2016 12:05:55 Stephan Kulow wrote:
> >      send_key "alt-f4";
> > 
> > -    sleep 1;
> > 
> >      send_key "alt-f4";
> 
> I was more thinking along
> ```
> wait_screen_change {
>     send_key "alt-f4";
> }
> send_key "alt-f4";

sometimes … you seem to have a bad timing ;-) changed as you proposed, works, 
too.

aaannz commented on tests/x11/sshxterm.pm:
```diff
@@ -16,19 +16,22 @@ sub run() {
-    type_string "yes\n";
-    wait_idle 6;
+    my $ret = check_screen([qw/ssh-xterm-host-key-authentication ssh-password-prompt/]);
+    die "ssh did not respond in time or needle(s) not found" unless $ret;
```
Why the `check_screen` and `die` combination and not `assert_screen`?

aaannz commented on tests/x11/sshxterm.pm:
```diff
@@ -16,19 +16,22 @@ sub run() {
+    my $ret = check_screen([qw/ssh-xterm-host-key-authentication ssh-password-prompt/]);
+    die "ssh did not respond in time or needle(s) not found" unless $ret;
+    # if ssh asks for authentication of the key accept it
+    if ($ret->{needle}->has_tag('ssh-xterm-host-key-authentication')) {
```
There is a handy helper `testapi::match_has_tag`. You then don't need to store matched needle, just call `match_has_tag('ssh-xterm-host-key-authentication');`

coolo replied:
it would be good to port our tests to use that - because we have plenty of such old style constructs around to copy&paste from ;(

okurz replied:
fixed with #968

okurz commented:
Thanks I will try all of your suggestions.

okurz commented:
updated:
* replaced check_screen with assert_screen
* use "match_has_tag"
* safeguard close window with "wait_screen_change"

```perl
-    my $ret = check_screen([qw/ssh-xterm-host-key-authentication ssh-password-prompt/]);
-    die "ssh did not respond in time or needle(s) not found" unless $ret;
+    assert_screen([qw/ssh-xterm-host-key-authentication ssh-password-prompt/]);
     # if ssh asks for authentication of the key accept it
-    if ($ret->{needle}->has_tag('ssh-xterm-host-key-authentication')) {
+    if (match_has_tag('ssh-xterm-host-key-authentication')) {
@@ -31,7 +30,10 @@ sub run() {
-    send_key "alt-f4";
+    # make sure the previous window closed
+    wait_screen_change {
+        send_key "alt-f4";
+    }
```

aaannz commented:
:+1:

coolo commented on tests/x11/sshxterm.pm:
```diff
@@ -16,19 +16,24 @@ sub run() {
+    # make sure the previous window closed
+    wait_screen_change {
+        send_key "alt-f4";
+    }
```
you're missing a semicolon here - I wonder this worked for you
